test/e2e/replica_set_x509/replica_set_x509_test.go

```diff
@@ -0,0 +1,222 @@
+package replica_set_x509
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "github.com/mongodb/mongodb-kubernetes-operator/api/v1"
+	"github.com/mongodb/mongodb-kubernetes-operator/pkg/util/constants"
+	e2eutil "github.com/mongodb/mongodb-kubernetes-operator/test/e2e"
+	"github.com/mongodb/mongodb-kubernetes-operator/test/e2e/mongodbtests"
+	"github.com/mongodb/mongodb-kubernetes-operator/test/e2e/setup"
+	"github.com/mongodb/mongodb-kubernetes-operator/test/e2e/tlstests"
+	. "github.com/mongodb/mongodb-kubernetes-operator/test/e2e/util/mongotester"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMain(m *testing.M) {
+	code, err := e2eutil.RunTest(m)
+	if err != nil {
+		fmt.Println(err)
+	}
+	os.Exit(code)
+}
+
+func TestReplicaSetX509(t *testing.T) {
+	resourceName := "mdb-tls"
+	helmArgs := []setup.HelmArg{
+		{Name: "resource.tls.useX509", Value: "true"},
+		{Name: "resource.tls.sampleX509User", Value: "true"},
+	}
+	ctx, testConfig := setup.SetupWithTLS(t, resourceName, helmArgs...)
+	defer ctx.Teardown()
+
+	mdb, _ := e2eutil.NewTestMongoDB(ctx, resourceName, testConfig.Namespace)
+	mdb.Spec.Security.Authentication.Modes = []v1.AuthMode{"X509"}
+	mdb.Spec.Security.TLS = e2eutil.NewTestTLSConfig(false)
+
+	tester, err := FromX509Resource(t, mdb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("User with wrong certificates", func(t *testing.T) {
+		mdb.Spec.Users = []v1.MongoDBUser{
+			getInvalidUser(),
+		}
+		users := mdb.GetAuthUsers()
+
+		t.Run("Create MongoDB Resource", mongodbtests.CreateMongoDBResource(&mdb, ctx))
+		t.Run("Basic tests", mongodbtests.BasicFunctionalityX509(&mdb))
+		t.Run("Agent certificate secrets configured", mongodbtests.AgentX509SecretsExists(&mdb))
+
+		cert, root, dir := createCerts(t, &mdb)
```

Comment:
why do these certs fail and below succeed?

Reply:
Because the certs that createCerts() creates are valid only for the user created by getValidUser(). In this case, the resource only has getInvalidUser(). So here, it is trying to connect with the certificates of a non-existent user.

Reply:
Oh i see, in that case I would call the test `wrong user with certificates` or similar? I kind of overlooked the beginning of the test